hibernate-core/src/main/java/org/hibernate/engine/internal/EntityEntryContext.java

```diff
@@ -107,21 +110,19 @@ public void addEntityEntry(Object entity, EntityEntry entityEntry) {
 				else {
 					// Create a holder for PersistenceContext-related data.
 					managedEntity = new ImmutableManagedEntityHolder( managed );
-					if ( immutableManagedEntityXref == null ) {
-						immutableManagedEntityXref = new IdentityHashMap<>();
+					if ( !isReferenceCachingEnabled( entityEntry.getPersister() ) ) {
+						managed.$$_hibernate_setInstanceId( nextManagedEntityInstanceId() );
```

Comment:
```suggestion
						assert managed.$$_hibernate_getInstanceId() == -1;
						managed.$$_hibernate_setInstanceId( nextManagedEntityInstanceId() );
```

Reply:
The entity might have been associated to a different persistence context, we don't reset instance ids.